assets/app/scripts/services/metrics.js

```diff
@@ -0,0 +1,122 @@
+'use strict';
+
+angular.module("openshiftConsole")
+  .factory("MetricsService", function($http, $q, APIDiscovery, Logger) {
+    var COUNTER_TEMPLATE = "/counters/{containerName}%2F{podUID}%2F{metric}/data";
+    var GAUGE_TEMPLATE = "/gauges/{containerName}%2F{podUID}%2F{metric}/data";
+    var RATE_TEMPLATE = "/counters/{containerName}%2F{podUID}%2F{metric}/rate";
+
+    // URL template to show for each type of metric.
+    var templateByMetric = {
+      "cpu/usage": COUNTER_TEMPLATE,
+      "memory/usage": GAUGE_TEMPLATE,
+      "memory/page_faults": RATE_TEMPLATE,
+      "memory/major_page_faults": RATE_TEMPLATE,
+      uptime: RATE_TEMPLATE
+    };
+
+    var metricsURL;
+    function getMetricsURL() {
+      if (angular.isDefined(metricsURL)) {
+        return $q.when(metricsURL);
+      }
+
+      return APIDiscovery.getMetricsURL().then(function(url) {
+        // Remove trailing slash if present.
+        metricsURL = (url || '').replace(/\/$/, "");
+        return metricsURL;
+      });
+    }
+
+    function normalizeMetricsData(data, metric) {
+      // Track the previous value for CPU usage calculations.
+      var lastValue;
+      if (!data.length) {
+        return;
+      }
+
+      angular.forEach(data, function(point) {
+        // Calculate a timestamp based on the midtime if missing.
+        if (!point.timestamp) {
+          point.timestamp = point.start + (point.end - point.start) / 2;
+        }
+
+        // Set point.value to the average or 0 if missing.
+        if (!point.value || point.value === "NaN") {
+          point.value = point.avg || 0;
+        }
+
+        // Convert cumulative CPU usage in nanoseconds to millicores.
+        if (metric === 'cpu/usage') {
+          var value = point.value;
+          if (angular.isDefined(lastValue)) {
+            var timeInMillis = point.end - point.start;
+            // Find the usage for just this bucket by comparing it to the last value.
+            // Values are in nanoseconds. Calculate usage in millis.
+            var usageInMillis = (value - lastValue) / 1000000;
+            // Convert to millicores.
+            point.value = (usageInMillis / timeInMillis) * 1000;
+          }
+
+          lastValue = value;
+        }
+      });
+
+      // Remove the first value since it can't be used CPU utilization.
+      // We want the same number of data points for all charts.
+      data.shift();
```

Comment:
should we ask for the data point prior to the time range we asked for so that we dont drop the data point at start?